pkg/asset/agent/mirror/registriesconf.go

```diff
@@ -208,6 +214,20 @@ func (i *RegistriesConf) finish() error {
 	return nil
 }
 
+func (i *RegistriesConf) validateReleaseImageIsSameInRegistriesConf() error {
+
+	releaseImage := &releaseimage.Image{}
+	releaseImage.Generate(asset.Parents{})
+
+	releaseImagePath := strings.Split(releaseImage.PullSpec, ":")[0]
+
+	if !strings.Contains(string(i.File.Data), releaseImagePath) {
```

Comment:
@bfournie is what we're looking for here really that [getMirrorFromRelease()](https://github.com/openshift/installer/blob/master/pkg/asset/agent/image/ignition.go#L381-L398) returns a non-empty value when a mirror config is set? I worry that the logic here is too simplistic.

Also is it really forbidden to add an ICSP that doesn't attempt to set up a mirror for the release? The user could just want to add an ICSP for their own images in the cluster on day 1, couldn't they? (Admittedly this seems unusual unless they are disconnected, in which case the release mirror really is needed.) Should this be a warning instead?

Reply:
What's the reason to perform the validation against a string instead of working with a `sysregistriesv2.V2RegistriesConf` type? It would be easier to validate a struct. Moreover, I'd expect that the unmarshaling would happen in the load (a common habit for those assets that are loading manifests from the disk), also to immediately verify that a valid `registries.conf` has been passed.
This will make easier to apply the validations in the `finish()` method, and it will avoid specifying the source which I found somehow complex to manage: ie, the validations should be the same regardless of how the entity has been fetched, loaded or generated.

Reply:
Just to be clear, I think we should always follow this common pattern in all of our writable assets producing a type T:

* **Generate**:  create `t T` from InstallConfig(+AgentConfig) if present, otherwise provide a reasonable default. Then validate via `finish(t)`
* **Load**: create `t T` by unmarshaling the file if present, then validate via `finish(t)` (otherwise do nothing)

Reply:
Currently, the unmarshalling is done in `extractLocationMirrorDataFromRegistries` from `finish()` which is called from `Generate()` as well as `Load()`. @bfournie Can you suggest why the registriesconf asset is handled a bit different here? All the other assets follow a different approach i.e. `finish()` only performs the validations and `Load()` does the unmarshalling.

@andfasano suggesting here to refactor this code such that all the assets follows the same approach.

Reply:
@andfasano I would like to do the refactoring in a separate PR

Reply:
I agree with Andrea that it's deeply weird that we're just searching for the release string somewhere in the file without reference to the structure of it.
I still don't think I agree with Bob that we should prevent the user moving forward if they have mirrors but not of the release image. I think this should be a warning, not an error.

Reply:
@pawanpinjarkar personally I don't see any valid reason to postpone such change, especially because it will greatly improve the readability and code maintenance by making this asset more similar to the other ones, with a few lines changed.
I'd also appreciate an improved test coverage on invalid config file, which apparently seems to be missing.